src/python/pants/engine/process.py

```diff
@@ -367,9 +368,15 @@ async def find_binary(request: BinaryPathRequest) -> BinaryPaths:
         fi
         """
     )
-    script_digest = await Get(
-        Digest,
-        CreateDigest([FileContent(script_path, script_content.encode(), is_executable=True)]),
+
+    # We get a UUID so that we ignore the cache every time, as this script depends on the state of
+    # the external environment.
+    script_digest, uuid = await MultiGet(
```

Comment:
IIRC this will cause every rule that depends on this product to rerun every time. Is that really what we want?

Reply:
I think if the end result of the rule is the same, those downstream rules won't be invalidated. It only means that this Param will always be re-evaluated. This is the same with how the new TestEnvironment type works. 

@gshuflin can you please confirm this is correct?

Reply:
I believe that the uncacheable UUID Get here will only cause this rule to have to rerun, rather than invalidating every downstream rule.

Reply:
Sure, but every *upstream* rule will have to rerun. Is that what we want? What uses `find_binary`?

Reply:
>  but every upstream rule will have to rerun. Is that what we want? What uses find_binary?

If your Python interpreters changed? Yes, absolutely, I think we should rerun every dependent rule.

For example, if you removed Python 3.7 from your machine and installed Python 3.8, that should invalidate your Python processes. It's not guaranteed that your tests would still pass.

Otherwise, the only way to get Pants to re-search for your interpreters is to purge `lmdb_store`, or modify `--python-setup-interpreter-search-paths`. There's no other way to get Pants to re-evaluate where your interpreters are.

--

Atm, `find_binary` is only used for finding Python. In the example plugin repo, it's used to find `zip` and `bash`.

Reply:
This won't only rerun if your python interpreters changed, it will rerun every single time. So basically anything that depends on a Python interpreter will rerun on every Pants invocation. Or at least so is my understanding.

Reply:
> it will rerun every single time

This specific rule will rerun every time, but the downstream rules will only rerun if the output of this specific rule changes.

This is the same with Greg's new `--test-extra-env` feature he added. That re-evaluated `os.environ` every single Pants run, yet we are still able to get caching.

Reply:
What happens if a rule `await Get`s on this rule (or something that depends on it)? The entire body of that rule has to rerun every time as well, no? Since this rule might be run conditionally.

Reply:
A not-so-side-note: Fixing things for the Pants client side is great, but we never have this problem for remoting since we effectively have a hash of the remote executuion image as a key in Procss executions via remote_execution_extra_platform_properties - `"container-image=..."`. So if the fix affects both sides its good to keep in mind we could probably strive to do better in the future and not impact remoting somehow with an evolution of the current client side fix.

Reply:
> This specific rule will rerun every time, but the downstream rules will only rerun if the output of this specific rule changes.

This is correct. Uncacheable nodes are not completely free, but they are cheaper than actually re-running the logic above the uncacheable node.

An uncacheable node runs once per "session" (typically one pants run), and everything that depends on an uncacheable node is effectively always marked "dirty". Dirty nodes do not re-run unless their dependencies (recursively, all the way down to the uncacheable node) have changed output values, represented using an integer generation value. That rust-only recursion on integer graph ids and integer generation values is much cheaper than actually running the nodes.

BUT: two things.
1. It's possible that this patch will experience the kinds of issues that @gshuflin ran into on #10129 (...although we've been using uncacheable nodes for `test --force`, so perhaps not)
2. This might be better modeled (eventually?) as a native operation of the `CommandRunner`s as I suggested originally on #9760: see the followup ticket I created after #9760 landed: #10526.

Reply:
👍